README.md
<div align="center">

# MeowDB.js
![MeowDB](https://i.imgur.com/cC7AZ18.png)

![Downloads](https://img.shields.io/npm/dt/meowdb)  ![Minified Size](https://img.shields.io/bundlephobia/min/meowdb) ![Vulnerabilities](https://img.shields.io/snyk/vulnerabilities/npm/meowdb) ![License](https://img.shields.io/npm/l/meowdb) ![Last Commit](https://img.shields.io/github/last-commit/Drylozu/MeowDB.js) ![GitHub Repo stars](https://img.shields.io/github/stars/Drylozu/MeowDB.js?style=social)

[![NPM](https://nodei.co/npm/meowdb.png?downloads=true&downloadRank=true&stars=true)](https://nodei.co/npm/meowdb/)
</div>

"Database" in JSON (Node.JS Library).

**Released v2.2.3**. See [CHANGELOG](https://github.com/Drylozu/MeowDB.js/blob/master/CHANGELOG.md).


## Installation
- `npm install meowdb --save`.

_I recommend not using versions lower than **2.1.9** or being aware of updates to the library._


## Usage
**JavaScript - CommonJS require**
```js
const MeowDB = require("meowdb");

const myDatabase = new MeowDB({
    dir: __dirname,
    name: "database",
    raw: false // Defines if MeowDBObjects will be returned (optional, default: false)
});
```

**TypeScript - ES6 import**

With TypeScript you should've the `esModuleInterop` flag.
```ts
import MeowDB from "meowdb";
// The generic type is optional, by default it's "full" but when using the raw option, use "raw" instead of "full"
const myDatabase = new MeowDB<'full'>({
    dir: __dirname,
    name: "database",
    raw: false // Defines if MeowDBObjects will be returned (optional, default: false)
});
```

### Example of all functions
```js
// Creating object (it'll search property by property and if it doesn't exist, it'll create it otherwise it'll not modify the current information~)
// * where the first parameter is the ID, they're like properties of an object (same thing in most functions)
const newObject = myDatabase.create("0001", {
    name: "David",
    country: "CO",
    info: "Nothing to show"
});
console.log(newObject);

// Obtaining an object
const object = myDatabase.get("0001");
console.log(object);

// Modifying an object and saving it
object.name = "Deivid";
object.save();
console.log(object.name);

// Setting directly the value of an element
const newName = myDatabase.set("0001.info", "Just a person");
console.log(newName);

// Listing all objects
let temp = "";
Object.entries(myDatabase.all()).forEach((user) => {
    temp += `   - ${user[1].name} (ID: ${user[0]})\n`;
});
console.log(temp.trimRight());

// Finding an object
const anObject = myDatabase.find((user) => user.name === "Deivid");
console.log(anObject);

// Filtering objects
const someObjects = myDatabase.filter((user) => user.country === "CO");
console.log(someObjects);

// Deleting an object
const deletedObject = myDatabase.delete("0001");
console.log(deletedObject);
```

### Important note while using TypeScript
You can use TypeScript Generics to `create`/`get`/`update`/`set`/`find`/`filter` the data, it doesn't matter what type you use.
```ts
const nonObjectValue = myDatabase.get<string>('0002.name');
console.log(nonObjectValue); // TS will interpret it as string

const numberValue = myDatabase.get<number>('some id here');
console.log(numberValue); // TS will interpret it as a number

const booleanValue = myDatabase.get<boolean>('some id here');
console.log(booleanValue); // TS will interpret it as a boolean

// With Objects/also works with interfaces
type Person = {
    name: string;
    country: string;
    info: string;
};

const objectValue = myDatabase.get<Person>('0002'); // This will return a MeowDBObject with the properties that you specified in the generic type
console.log(objectValue.name); // While typing '.name', you'll get *autocomplete*
// It also works when you save an MeowDBObject
objectValue.info = 'Hi!';

/// Important: Read the note in the Usage/TypeScript section.
objectValue.save(); // This will return a plain 'Person' object.
```

## "Documentation"
- [`new MeowDB(options)`](#new-meowdboptions)
    - `create(id, initialValue)`
    - `exists(id)`
    - `get(id)`
    - `set(id, value)`
    - `all()`
    - `delete(id)`
    - `find(callback, id?)`
    - `filter(callback, id?)`
- [`MeowDBError`](#meowdberror)


## new MeowDB(options)
Creates or gets a database
- **Parameters**:
    - `options` - An object with the options
        - `options.dir` - A string indicating the directory that will have the database (must be an absolute path - the folder should be created)
        - `options.name` - A string with the name of the database
        - `options.raw?` - A boolean that represents if MeowDBObjects won't returned (default: `false`, MeowDBObjects will be returned).
- **Throws**: [`MeowDBError`](#meowdberror) - If any option is invalid


### Methods
#### `all()`
Returns all data stored in the database
- **Returns**: `MeowDBObject | object` - All data
<hr>

#### `create(id, initialValue)`
Creates an element in the database with the specified ID and sets it's value
- **Parameters**:
    - `id` - A string representing the ID of the element to create
    - `initialValue` - The initial value of the element
- **Returns**: `object` - The created element
- **Throws**: [`MeowDBError`](#meowdberror) - If the ID or initialValue is invalid
<hr>

#### `delete(id)`
Deletes an element from the database
- **Parameters**:
    - `id` - A string representing the ID of the element to delete
- **Returns**: `object` - The deleted element
- **Throws**: [`MeowDBError`](#meowdberror) - If the ID is invalid
<hr>

#### `exists(id)`
Checks if an element exists in the database
- **Parameters**:
    - `id` - A string representing the ID of the element to check
- **Returns**: `boolean` - If it exists
- **Throws**: [`MeowDBError`](#meowdberror) - If the ID is invalid
<hr>

#### `get(id)`
Gets an element of the database
- **Parameters**:
    - `id` - A string representing the ID of the element to get
- **Returns**: `MeowDBObject | object | any` - The element
- **Throws**: [`MeowDBError`](#meowdberror) - If the ID is invalid
<hr>

#### `set(id, value)`
Sets the value of an element in the database
- **Parameters**:
    - `id` - A string representing the ID of the element to update
    - `value` - The new value of the element
- **Returns**: `any` - The value setted
- **Throws**: [`MeowDBError`](#meowdberror) - If the ID or value is invalid
<hr>

#### `find(callback, id?)`
Finds an element in the database.
You __should only__ use this function if you're finding for objects
- **Parameters**:
- `callback` - A function that handles all the elements and decides which one will be returned
    - `id?` - A string representing the ID of the root element to find another elements _(optional)_
- **Returns**: `MeowDBObject | object | any` - The element
- **Throws**: [`MeowDBError`](#meowdberror) - If the ID or callback is invalid
<hr>

#### `filter(callback, id?)`
Filters elements in the database.
You __should only__ use this function if you're filtering for objects
- **Parameters**:
    - `callback` - A function that handles all the elements and decides which ones will be returned
    - `id?` - A string representing the ID of the root element to find another elements _(optional)_
- **Returns**: `(MeowDBObject | object | [string, any])[]` - The elements (MeowDBObject[] if they're objects, array with ID and value if not)
- **Throws**: [`MeowDBError`](#meowdberror) - If the ID or callback is invalid


## MeowDBError
Extends [`Error`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error), only used for error reference.

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## 2.2.3
- Added `raw` property to the MeowDB constructor options, to don't instantiate a `MeowDBObject` class every time.

## 2.2.2 - More TypeScript!!!
- You can use TypeScript Generics with MeowDB now. [README](https://github.com/Drylozu/MeowDB.js/blob/master/README.md#important-note-while-using-typescript).

## 2.2.1 - yey
- Better ID handling.
- Improved TS typings (callbacks).
- Improved MeowDBObject with `__id` and `__file` properties.
- Added `.npmignore` to exclude some files in NPM.
- Improved ESLint.
- When filtering or finding, the generated MeowDBObject didn't save the full ID of the object, so the function `save` wouldn't save well.
- Improved JSDocs.
- Some functions are now *static* (in `MeowDBUtils`).

## 2.2.0 - Improved MeowDB in general
- Removed all TS things (only .d.ts file).
- Removed dependencies badge (buggy).
- Improved README.
- Improved error messages.
- Improved TS typings (now, they work).

## 2.1.9 - Fixed some bugs with some symbols
- IDs and values (string) now accepts every character.
- Bugs with IDs and values fixed (security errors).
_Be free to use MeowDB now :)_

## 2.1.8 - Improved TS typings
- Updated dependencies. 
- Modified TS declaration file.

## 2.1.7 - Improved data validation
- Updated devDependencies.
- Now all characters are accepted except line breaks (`\n`) in IDs (yes.).
- Now undefined is returned if the value isn't supported.

## 2.1.6 - Updates
- Updated devDependencies.

## 2.1.5 - More functions
- Added `find` and `filter` functions.
- `README.md` more beautiful.
- The ID of an element now supports breaklines (`\n`).

## 2.1.4 - Error fixed
- `push` in an array (MeowDBObject property) works now.

## 2.1.3 - Better errors
- MeowDB functions don't returns an error now.
- MeowDB functions throws an error now.
- MeowDBObject now uses `Object.assign`.
- Removed `-` in JSDocs.
- Updated version in README

## 2.1.2 - JSDocs
- Added JSDoc in TS declaration file.
- Added CHANGELOG file.
- Changed MeowDB Banner from README.
- Added use of `set` function in the example.

## 2.1.1 - Added TS support
- Updated LICENSE file.
- Added MeowDB Banner in README.
- Added TS Import in README example.
- Added TS declaration file (for the TS support).
- Added JSDoc in every JS file.

## 2.1.0 - Unpromisified
- Added `eslint` to the project.
- Added `.gitignore` (before there was nothing to ignore, now there's).
- Added shields.io images in README.
- Restructured example.
- Renamed project from `meowdb` to `MeowDB.js`.
- Added Ruby repository/gem to README.

## 2.0.7 - Error fixed
- `set(id, array)` didn't work well before, now it does (It threw an error).

## 2.0.6 - Unpromisified function
- `get` function don't returns a Promise now.

## 2.0.5 - More private things
- `_options` and `_utils` properties are now private in the MeowDB class.
- Added `@returns` (JSDoc) in every function.

## 2.0.4 - A "big" error
- `sValue` -> `value` in `set` function.

## 2.0.3 - Little things fixed
- Changed actual version from v2.0.1 to v2.0.3 in the README file.

## 2.0.2 - Another little things
- Actual version ("*Released v-.-.-*") is now in **bold**.

## 2.0.1 - Spelling errors fixed
- Some things added/removed from the README.

## 2.0.0 - Rework
- Restructured README, less information and more facility.
- Removed every file/class/function (without lossing his structure).
- `checkName` is now in the MeowDB constructor.
- Synchronous functions but returns `Promise`.
- The utils are now a class.

## 1.2.0 - More information
- Added example in README.
- Added a long documentation for all files in README.
- More code in the example file.

## 1.1.0 - Private things
- Renamed `structures/` classes.
- Functions work independently.
- `checkId` function now supports *all* (*).
- `createData` function without an default value in `sValue` parameter.

## 1.0.1 - Spelling errors fixed
- Added `p` in `example` (test script of `package.json`).

## 1.0.0 - First release
- Added main files (LICENSE, README, structures/ and index).
- JSDocs in JS files.
- Create, delete, get, set, all and sort functions.
- **A bad thing**, everything async.

example/index.ts
console.time("MeowDB.js");

import MeowDB from "../";

const db: MeowDB = new MeowDB({
    dir: __dirname,
    name: "test"
});

console.log("Object creation (only if it doesn't exist)");
console.log(db.create("0001", {
    name: "David",
    country: "CO",
    info: "Nothing to show"
}), "\n");

console.log("Object creation (only if it doesn't exist)");
console.log(db.create("0002", {
    name: "Free",
    country: "NI",
    info: "Nothing to show"
}), "\n");

console.log("Obtaining an object with ID 0001");
let object = db.get("0001");
console.log(object, "\n");

console.log("Editing 'name' key and saving");
process.stdout.write(`${object.name} - `);
object.name = "Deivid";
object.save();
console.log(object.name, "\n");

console.log("Setting directly the value of an element");
console.log(db.set("0002.info", "Just a person"), "\n");

console.log("List of objects");
let temp = "";
Object.entries(db.all()).forEach((u: [string, any]) => {
    temp += `   - ${u[1].name} (${u[0]})\n`;
});
console.log(temp.trimRight(), "\n");

console.log("Finding an object with property 'country' as 'NI'");
console.log(db.find((u: any) => u.country === "NI"), "\n");

console.log("Filtering objects with property 'name'");
console.log(db.filter((u: any) => u.name), "\n");

console.log("Deleting an object");
console.log(db.delete("0001"), "\n");

console.timeEnd("MeowDB.js");